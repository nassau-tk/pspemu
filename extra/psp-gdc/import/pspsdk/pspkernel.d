module pspsdk.pspkernel;

/*
#include <pspuser.h>
#include <pspiofilemgr_kernel.h>
#include <psploadcore.h>
#include <pspstdio_kernel.h>
#include <pspsysreg.h>
#include <pspkdebug.h>
#include <pspintrman_kernel.h>
#include <pspmodulemgr_kernel.h>
*/

/**
 * Set the $pc register to a kernel memory address.
 *
 * When the PSP's kernel library stubs are called, they expect to be accessed
 * from the kernel's address space.  Use this function to set $pc to the kernel
 * address space, before calling a kernel library stub.
 */
/*
#define pspKernelSetKernelPC()  \
{     \
	__asm__ volatile (      \
	"la     $8, 1f\n\t"     \
	"lui    $9, 0x8000\n\t" \
	"or     $8, $9\n\t"     \
	"jr     $8\n\t"         \
	" nop\n\t"              \
	"1:\n\t"                \
	: : : "$8", "$9");      \
	sceKernelIcacheClearAll(); \
}
*/

/**
 * Set the $pc register to a user memory address.
 *
 */
/*
#define pspKernelSetUserPC()  \
{     \
	__asm__ volatile (      \
	"la     $8, 1f\n\t"     \
	"li     $9, 0x7FFFFFFF\n\t" \
	"and    $8, $9\n\t"     \
	"jr     $8\n\t"         \
	" nop\n\t"              \
	"1:\n\t"                \
	: : : "$8", "$9");      \
	sceKernelIcacheClearAll(); \
}
*/