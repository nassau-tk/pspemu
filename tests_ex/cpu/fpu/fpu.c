/*
FPU Test. Originally from jpcsp project:
http://code.google.com/p/jpcsp/source/browse/trunk/demos/src/fputest/main.c
Modified to perform automated tests.
*/

#include <pspkernel.h>

PSP_MODULE_INFO("fpu test", 0, 1, 1);
PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER);

float __attribute__((noinline)) adds(float x, float y) {
	float result;
	asm volatile("add.s %0, %1, %2" : "=f"(result) : "f"(x), "f"(y));
	return result;
}

float __attribute__((noinline)) subs(float x, float y) {
	float result;
	asm volatile("sub.s %0, %1, %2" : "=f"(result) : "f"(x), "f"(y));
	return result;
}

float __attribute__((noinline)) muls(float x, float y) {
	float result;
	asm volatile("mul.s %0, %1, %2" : "=f"(result) : "f"(x), "f"(y));
	return result;
}

float __attribute__((noinline)) divs(float x, float y) {
	float result;
	asm volatile("div.s %0, %1, %2" : "=f"(result) : "f"(x), "f"(y));
	return result;
}

float __attribute__((noinline)) sqrts(float x) {
	float result;
	asm volatile("sqrt.s %0, %1" : "=f"(result) : "f"(x));
	return result;
}

float __attribute__((noinline)) abss(float x) {
	float result;
	asm volatile("abs.s %0, %1" : "=f"(result) : "f"(x));
	return result;
}

float __attribute__((noinline)) negs(float x) {
	float result;
	asm volatile("neg.s %0, %1" : "=f"(result) : "f"(x));
	return result;
}

int __attribute__((noinline)) cvtws(float x, int rm) {
	float resultFloat;
	asm volatile("ctc1 %0, $31" : : "r"(rm));
	asm volatile("cvt.w.s %0, %1" : "=f"(resultFloat) : "f"(x));
	int result = *((int *) &resultFloat);
	return result;
}

#define CHECK_OP(op, expected) { float f = 0.0; f = op; Kprintf("%s\n%f\n", #op " == " #expected, f);}

#define RINT_0  0
#define CAST_1  1
#define CEIL_2  2
#define FLOOR_3 3

int main(int argc, char *argv[]) {
	CHECK_OP(adds(1.0, 1.0), 2.0);
	CHECK_OP(subs(3.0, 1.0), 2.0);
	CHECK_OP(muls(2.0, 1.0), 2.0);
	CHECK_OP(divs(4.0, 2.0), 2.0);
	CHECK_OP(abss(+2.0), 2.0);
	CHECK_OP(abss(-2.0), 2.0);
	CHECK_OP(negs(negs(+2.0)), 2.0);
	CHECK_OP(sqrts(4.0), 2.0);

	CHECK_OP(cvtws(1.1, RINT_0), 1);
	CHECK_OP(cvtws(1.1, CAST_1), 1);
	CHECK_OP(cvtws(1.1, CEIL_2), 2);
	CHECK_OP(cvtws(1.1, FLOOR_3), 1);

	CHECK_OP(cvtws(-1.1, RINT_0), -1);
	CHECK_OP(cvtws(-1.1, CAST_1), -1);
	CHECK_OP(cvtws(-1.1, CEIL_2), -1);
	CHECK_OP(cvtws(-1.1, FLOOR_3), -2);

	CHECK_OP(cvtws(1.9, RINT_0), 2);
	CHECK_OP(cvtws(1.9, CAST_1), 1);
	CHECK_OP(cvtws(1.9, CEIL_2), 2);
	CHECK_OP(cvtws(1.9, FLOOR_3), 1);

	CHECK_OP(cvtws(1.5, RINT_0), 2);
	CHECK_OP(cvtws(1.5, CAST_1), 1);
	CHECK_OP(cvtws(1.5, CEIL_2), 2);
	CHECK_OP(cvtws(1.5, FLOOR_3), 1);

	return 0;
}
