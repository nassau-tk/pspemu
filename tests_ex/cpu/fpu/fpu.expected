adds(1.0, 1.0) == 2.0
2.000000
subs(3.0, 1.0) == 2.0
2.000000
muls(2.0, 1.0) == 2.0
2.000000
divs(4.0, 2.0) == 2.0
2.000000
abss(+2.0) == 2.0
2.000000
abss(-2.0) == 2.0
2.000000
negs(negs(+2.0)) == 2.0
2.000000
sqrts(4.0) == 2.0
2.000000
cvtws(1.1, RINT_0) == 1
1.000000
cvtws(1.1, CAST_1) == 1
1.000000
cvtws(1.1, CEIL_2) == 2
2.000000
cvtws(1.1, FLOOR_3) == 1
1.000000
cvtws(-1.1, RINT_0) == -1
-1.000000
cvtws(-1.1, CAST_1) == -1
-1.000000
cvtws(-1.1, CEIL_2) == -1
-1.000000
cvtws(-1.1, FLOOR_3) == -2
-2.000000
cvtws(1.9, RINT_0) == 2
2.000000
cvtws(1.9, CAST_1) == 1
1.000000
cvtws(1.9, CEIL_2) == 2
2.000000
cvtws(1.9, FLOOR_3) == 1
1.000000
cvtws(1.5, RINT_0) == 2
2.000000
cvtws(1.5, CAST_1) == 1
1.000000
cvtws(1.5, CEIL_2) == 2
2.000000
cvtws(1.5, FLOOR_3) == 1
1.000000